server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala

```diff
@@ -199,9 +199,20 @@ private[deploy] class Worker(
       HeartbeatFromWorker(host, rpcPort, pushPort, fetchPort, replicatePort, workerInfo.numSlots,
         shuffleKeys)
       , classOf[HeartbeatResponse])
-    cleanTaskQueue.put(response.expiredShuffleKeys)
-    if (!response.registered) {
-      logError("Current worker not registered in master")
+    if (response.registered) {
+      cleanTaskQueue.put(response.expiredShuffleKeys)
+    } else {
+      logError("Worker not registered in master, clean all shuffle data and register again.")
+      // Clean all shuffle related metadata and data
+      cleanup(shuffleKeys)
+      try {
+        registerWithMaster()
```

Comment:
To avoid concurrent issue, I think we should add epoch in worker. Each time it re-registers with master, increment its epoch, and ignore any heartbeat response from previous epoch.